docs/patterns/fileuploads.rst~

```diff
@@ -0,0 +1,190 @@
+   are patient :)
+
+   Now let's look how that function works:
+
+   >>> secure_filename('../../../../home/username/.bashrc')
+   'home_username_.bashrc'
+
+Now one last thing is missing: the serving of the uploaded files.  As of
+Flask 0.5 we can use a function that does that for us::
+
+    from flask import send_from_directory
+
+    @app.route('/uploads/<filename>')
+    def uploaded_file(filename):
+        return send_from_directory(app.config['UPLOAD_FOLDER'],
+                                   filename)
+
+Alternatively you can register `uploaded_file` as `build_only` rule and
+use the :class:`~werkzeug.wsgi.SharedDataMiddleware`.  This also works with
+older versions of Flask::
+
+    from werkzeug import SharedDataMiddleware
+    app.add_url_rule('/uploads/<filename>', 'uploaded_file',
+                     build_only=True)
+    app.wsgi_app = SharedDataMiddleware(app.wsgi_app, {
+        '/uploads':  app.config['UPLOAD_FOLDER']
+    })
+
+If you now run the application everything should work as expected.
+
+
+Improving Uploads
+-----------------
+
+.. versionadded:: 0.6
+
+So how exactly does Flask handle uploads?  Well it will store them in the
+webserver's memory if the files are reasonable small otherwise in a
+temporary location (as returned by :func:`tempfile.gettempdir`).  But how
+do you specify the maximum file size after which an upload is aborted?  By
+default Flask will happily accept file uploads to an unlimited amount of
+memory, but you can limit that by setting the ``MAX_CONTENT_LENGTH``
+config key::
+
+    from flask import Flask, Request
+
+    app = Flask(__name__)
+    app.config['MAX_CONTENT_LENGTH'] = 16 * 1024 * 1024
+
+The code above will limited the maximum allowed payload to 16 megabytes.
+If a larger file is transmitted, Flask will raise an
+:exc:`~werkzeug.exceptions.RequestEntityTooLarge` exception.
+
+This feature was added in Flask 0.6 but can be achieved in older versions
+as well by subclassing the request object.  For more information on that
+consult the Werkzeug documentation on file handling.
+
+
+Upload Progress Bars
+--------------------
+
+A while ago many developers had the idea to read the incoming file in
+small chunks and store the upload progress in the database to be able to
+poll the progress with JavaScript from the client.  Long story short: the
+client asks the server every 5 seconds how much it has transmitted
+already.  Do you realize the irony?  The client is asking for something it
+should already know.
+
+Now there are better solutions to that work faster and more reliable.  The
+web changed a lot lately and you can use HTML5, Java, Silverlight or Flash
+to get a nicer uploading experience on the client side.  Look at the
+following libraries for some nice examples how to do that:
+
+-   `Plupload <http://www.plupload.com/>`_ - HTML5, Java, Flash
+-   `SWFUpload <http://www.swfupload.org/>`_ - Flash
+-   `JumpLoader <http://jumploader.com/>`_ - Java
```

Comment:
I think it's a very bad idea to suggest anything Flash-based here. Much worse, anything Java-based.
Some beginners will think that using Java for this is acceptable. It's not. And Flash is "meh", but not really necessary nowadays.